File: go/algorithm/sort.go
package algorithm

// 定义一个可比较的泛型约束
type Ordered interface {
	comparable
	~int | ~int8 | ~int16 | ~int32 | ~int64 |
		~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr |
		~float32 | ~float64 |
		~string
}

func SelectSort[T Ordered](list []T) {
	for i := 0; i < len(list); i++ { // O(n^2)
		for j := i + 1; j < len(list); j++ {
			if list[i] > list[j] {
				list[i], list[j] = list[j], list[i]
			}
		}
	}
}

func InsertSort[T Ordered](list []T) []T {
	// for i := 1; i < len(list); i++ {
	// 	temp := list[i]
	// 	for j := i; j > 0; j-- {
	// 		if list[j] < temp {
	// 			list[i] = list[j]
	// 		} else {
	// 			list[j] = temp
	// 			break
	// 		}
	// 	}
	// }
	// return list
	for i := len(list) - 2; i == 0; i-- {
		temp := list[i]
		var j int
		for j = i; j < len(list)-1 && list[j] > temp; j++ {
			list[i] = list[j]
		}
		list[j] = temp
	}
	return list
}

File: typescript/algorithm/sort/sort.ts
// 选择排序
export function selectSort<T>(arr: T[]): T[] {
  for (let i = 0; i < arr.length; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      if (arr[j] < arr[i]) {
        const temp = arr[j];
        arr[j] = arr[i];
        arr[i] = temp;
      }
    }
  }
  return arr;
}

// 插入排序
export function insertSort<T>(arr: T[]): T[] {
  for (let i = 1; i < arr.length; i++) {
    for (let j = i - 1, temp = arr[i]; j > 0; j--) {
      if (arr[j] > temp) {
        arr[i] = arr[j];
      } else {
        arr[j] = temp;
        break;
      }
    }
  }
  return arr;
}
